Add ObjectSelectorStreamProducer template and use it for RecoTrackViewRefSelector

makortel opened the pull request:
Here is an attempt to address @Dr15Jones' comment (https://github.com/cms-sw/cmssw/pull/9201#issuecomment-109273147) about ignored EDFilters (in tracking validation sequence). Basically the PR adds new `ObjectSelector(Stream)Producer` templates similar to current `ObjectSelector(Stream)` but being EDProducers (I'm not sure if the legacy-default is still "needed", I decided to include it only to have similar code structure as `ObjectSelector(Stream)`).

As a use case I added RecoTrackViewRefSelector (existing RecoTrack(Ref)Selectors take `vector<Track>` as an input, and I'll soon need anyway a version reading the input via `View`), and switched the tracking MC validation sequence to use it.

Tested in CMSSW_7_6_X_2015-06-24-2300, no changes expected in results.

@rovere @VinInn

makortel committed:
Add RecoTrackViewRefSelector

makortel committed:
Switch to RecoTrackViewRefSelector in MTV

makortel committed:
Add ObjectSelector(Stream)Producer

These EDProducer-variants are intended for cases where the event
filtering functionality will not be needed.

cmsbuild commented:
A new Pull Request was created by @makortel (Matti Kortelainen) for CMSSW_7_6_X.

Add ObjectSelectorStreamProducer template and use it for RecoTrackViewRefSelector

It involves the following packages:

CommonTools/RecoAlgos
CommonTools/UtilAlgos
DataFormats/TrackReco
Validation/RecoTrack


@cvuosalo, @monttj, @cmsbuild, @deguio, @slava77, @vadler, @danduggan can you please review it and eventually sign? Thanks.
@rappoccio, @ahinzmann, @abbiendi, @GiacomoSguazzoni, @jhgoh, @VinInn, @appeltel, @jdolen, @rovere, @wmtford, @gpetruc, @istaslis, @cerati, @dgulhan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on CommonTools/UtilAlgos/interface/ObjectSelectorProducer.h:
```diff
@@ -0,0 +1,68 @@
+ */
+template<typename Selector,
+         typename OutputCollection = typename ::helper::SelectedOutputCollectionTrait<typename Selector::collection>::type,
+         typename PostProcessor = ::helper::NullPostProcessor<OutputCollection, edm::EDProducer>,
```
Do we even want to leave a default edm::EDProducer here?
Can this be just "typename blah" ?

makortel replied:
I don't disagree on dropping edm::EDProcucer. I'll then remove this header and make ObjectSelectorStreamProducer the main template definition.

slava77 replied:
in case edm::one or global instances of this template will be needed, maybe we should keep the generic definition.
I just didn't want to have the edm::EDProducer to be the default argument in the template args

makortel replied:
Right, taking out the default values for the template parameters of ObjectSelectorProducer (as the stream/global/one-ones would anyway have to give them explicitly) could be the most clear option.

Even with stream being the default, the one/global cases could still be instantiated from it (the only "penalty" wrt. current situation would be the inclusion of ".../stream/EDProducer.h" header where it is not strictly needed).

Anyway, I'll do an update with default values for template parameters of ObjectSelectorProducer removed.

cmsbuild commented:
+1
Tested at: 1eb5b38bbcfaad5fb9250cca1a6ac6cb754d1060
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9904/9/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9904/9/summary.html

makortel committed:
Remove default values of template parameters

The stream/global/one-specializations will anyway have to give them
explicitly.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #9904 was updated. @cvuosalo, @monttj, @cmsbuild, @deguio, @slava77, @vadler, @danduggan can you please check and sign again.

cmsbuild commented:
+1
Tested at: dc7169a0372120d2aa9f4560a753345f678e5d4e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9904/21/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9904/21/summary.html

makortel committed:
Switch also btvTrackRefs to RecoTrackViewRefSelector

makortel commented:
I realized only now that also the `btvTrackRefs` could/should be changed to `RecoTrackViewRefSelector`.

cmsbuild commented:
Pull request #9904 was updated. @cvuosalo, @monttj, @cmsbuild, @deguio, @slava77, @vadler, @danduggan can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f7d5d15da628c24b89eff97a564cc9e84955b486
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9904/8462/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9904/8462/summary.html

cvuosalo commented:
+1

For #9904 f7d5d15da628c24b89eff97a564cc9e84955b486

Adding a new Add ObjectSelectorStreamProducer template. There should be no change in monitored quantities.

The code changes are satisfactory, and Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9904/8462/summary.html) against baseline CMSSW_7_6_X_2015-06-30-2300 show no significant differences, as expected.

deguio commented:
+1

monttj commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs  (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

davidlange6 commented:
+1
